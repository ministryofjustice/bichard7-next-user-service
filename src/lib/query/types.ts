export type DBResult<T> = T
export type TransformResult<T> = T
export type Transform<I, O> = (dbResult: DBResult<I>) => TransformResult<O>
export type QueryError = Error
export type OnError = (e: QueryError) => void

export enum QueryType {
  Any = "Any"
}
