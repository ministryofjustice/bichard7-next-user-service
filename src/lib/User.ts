export interface UserCredentials {
  emailAddress: string
  password: string
}

export type UserRole = "B7Allocator" | "B7ExceptionHandler"

export interface User {
  displayName: string
  emailAddress: string
  role: UserRole
}
