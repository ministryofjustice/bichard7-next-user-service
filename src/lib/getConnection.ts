import Database from "types/Database"
import config from "./config"
import createSingletonConnection from "./createSingletonConnection"

const getConnection = (): Database => {
  return createSingletonConnection("users-service-connection", config.database)
}

export default getConnection
