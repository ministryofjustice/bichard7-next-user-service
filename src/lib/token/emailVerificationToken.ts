import jwt from "jsonwebtoken"
import config from "lib/config"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

const verifyOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type EmailVerificationToken = string

export interface EmailVerificationTokenPayload {
  emailAddress: string
  verificationCode: string
}

export function decodeEmailVerificationToken(token: EmailVerificationToken): EmailVerificationTokenPayload {
  try {
    return jwt.verify(token, config.tokenSecret, verifyOptions) as EmailVerificationTokenPayload
  } catch (error) {
    return error
  }
}

export function generateEmailVerificationToken(payload: EmailVerificationTokenPayload): EmailVerificationToken {
  const options: jwt.SignOptions = {
    expiresIn: `${config.emailVerificationExpiresIn} minutes`,
    ...signOptions
  }

  try {
    return jwt.sign(payload, config.tokenSecret, options)
  } catch (error) {
    return error
  }
}
