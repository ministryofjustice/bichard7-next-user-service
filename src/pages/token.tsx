import ErrorSummary from "components/ErrorSummary"
import GridRow from "components/GridRow"
import Layout from "components/Layout"
import Head from "next/head"
import { GetServerSideProps } from "next"
import Auth from "@aws-amplify/auth"

interface Props {
  failure?: boolean
  token?: string
  tokenContents?: string
}

export const getServerSideProps: GetServerSideProps = async () => {
  const props: Props = {}

  try {
    const user = await Auth.currentAuthenticatedUser()
    const idToken = user.getSignInUserSession()?.getIdToken()
    props.token = idToken?.getJwtToken()
    props.tokenContents = JSON.stringify(idToken?.decodePayload(), null, 4)
  } catch (error) {
    console.log(error)
    props.failure = true
  }

  return { props }
}

const Token = ({ failure, token, tokenContents }: Props) => (
  <>
    <Head>
      <title>{"Sign in to Bichard 7"}</title>
    </Head>
    <Layout>
      {failure && (
        <GridRow>
          <ErrorSummary title="Failed to get token">{"There was a problem getting the token"}</ErrorSummary>
        </GridRow>
      )}

      {token && (
        <GridRow>
          <code style={{ wordWrap: "break-word" }}>{token}</code>
        </GridRow>
      )}

      {tokenContents && (
        <GridRow>
          <pre>{tokenContents}</pre>
        </GridRow>
      )}
    </Layout>
  </>
)

export default Token
