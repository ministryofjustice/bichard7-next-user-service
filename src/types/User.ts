import { UserGroupResult } from "./UserGroup"

interface User {
  id: number
  username: string
  exclusionList: string[]
  inclusionList: string[]
  endorsedBy: string
  orgServes: string
  forenames: string
  surname: string
  emailAddress: string
  groups: UserGroupResult[]
}

export default User
