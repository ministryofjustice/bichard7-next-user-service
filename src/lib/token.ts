import jwt from "jsonwebtoken"
import config from "lib/config"
import { UserGroup } from "lib/User"

const signOptions: jwt.SignOptions = {
  issuer: config.tokenIssuer
}

export type Token = string

export interface BichardTokenPayload {
  username: string
  exclusionList: string[]
  inclusionList: string[]
  forenames: string
  surname: string
  emailAddress: string
  groups: UserGroup[]
}

export interface EmailTokenPayload {
  emailAddress: string
  verificationCode: string
}

export function generateBichardToken(payload: BichardTokenPayload): Token {
  const options: jwt.SignOptions = {
    expiresIn: config.tokenExpiresIn,
    ...signOptions
  }

  return jwt.sign(payload, config.tokenSecret, options)
}

export function generateEmailToken(payload: EmailTokenPayload): Token {
  const options: jwt.SignOptions = {
    expiresIn: "3 hours",
    ...signOptions
  }

  return jwt.sign(payload, config.tokenSecret, options)
}
