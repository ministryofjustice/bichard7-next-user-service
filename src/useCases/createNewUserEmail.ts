import generateNewUserEmail from "emails/newUser"
import { EmailVerificationTokenPayload, generateEmailVerificationToken } from "lib/token/emailVerificationToken"
import EmailContent from "types/EmailContent"
import { isError, Result } from "types/Result"
import UserCreateDetails from "types/UserDetails"

export default (user: UserCreateDetails, verificationCode: string): Result<EmailContent> => {
  const payload: EmailVerificationTokenPayload = {
    emailAddress: user.emailAddress,
    verificationCode
  }

  const token = generateEmailVerificationToken(payload)
  if (isError(token)) {
    return token
  }

  const url = new URL("/login/new-password", "http://localhost:3000")
  url.searchParams.append("token", token)

  return generateNewUserEmail({ url: url.href, user })
}
