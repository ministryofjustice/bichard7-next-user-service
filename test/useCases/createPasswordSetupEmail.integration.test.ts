/* eslint-disable import/first */
jest.mock("lib/token/emailVerificationToken")

import User from "types/User"
import UserCreateDetails from "types/UserDetails"
import getConnection from "lib/getConnection"
import createUser from "useCases/createUser"
import createNewUserEmail from "useCases/createNewUserEmail"
import { isError } from "types/Result"
import initialiseUserPassword from "useCases/initialiseUserPassword"
import storePasswordResetCode from "useCases/storePasswordResetCode"
import { generateEmailVerificationToken } from "lib/token/emailVerificationToken"
import deleteDatabaseUser from "./deleteDatabaseUser"
import EmailContent from "types/EmailContent"

const connection = getConnection()

const setupUser = {
  username: "SetupUsername",
  emailAddress: "SetupEmailAddress",
  exclusionList: "SetupExclusionList",
  inclusionList: "SetupInclusionList",
  endorsedBy: "SetupEndorsedBy",
  orgServes: "SetupOrgServes",
  forenames: "SetupForenames",
  postalAddress: "SetupPostalAddress",
  postCode: "AS2 2SA",
  phoneNumber: "SetupPhoneNumber"
} as unknown as User

const verificationCode = "123456"

describe("AccountSetup", () => {
  beforeAll(async () => {
    await deleteDatabaseUser(connection, setupUser.username)
  })

  afterAll(() => {
    connection.$pool.end()
  })

  it("should generate the email subject and body to request user to setup password", async () => {
    const mockedGeneratePasswordResetToken = generateEmailVerificationToken as jest.MockedFunction<
      typeof generateEmailVerificationToken
    >
    mockedGeneratePasswordResetToken.mockReturnValue("DUMMY_TOKEN")

    const createUserDetails: UserCreateDetails = {
      username: setupUser.username,
      forenames: setupUser.forenames,
      emailAddress: setupUser.emailAddress,
      endorsedBy: setupUser.endorsedBy,
      surname: setupUser.surname,
      organisation: setupUser.orgServes,
      postCode: setupUser.postCode,
      phoneNumber: setupUser.phoneNumber,
      postalAddress: setupUser.postalAddress
    }

    const result = await createUser(connection, createUserDetails)
    expect(isError(result)).toBe(false)

    const passwordSetCodeResult = await storePasswordResetCode(
      connection,
      createUserDetails.emailAddress,
      verificationCode
    )
    expect(isError(passwordSetCodeResult)).toBe(false)

    const newUserEmailResult = createNewUserEmail(createUserDetails, verificationCode)
    expect(isError(newUserEmailResult)).toBe(false)

    const email = newUserEmailResult as EmailContent
    expect(email.subject).toMatchSnapshot()
    expect(email.text).toMatchSnapshot()
    expect(email.html).toMatchSnapshot()
  })

  it("should be able to setup a password using the details from the email", async () => {
    const result = await initialiseUserPassword(connection, setupUser.emailAddress, verificationCode, "NewPassword")
    expect(result).toBeUndefined()
  })

  it("should not be able to setup a password a second time using the same details", async () => {
    const secondResult = await initialiseUserPassword(
      connection,
      setupUser.emailAddress,
      verificationCode,
      "NewPassword"
    )
    expect(secondResult).toBeDefined()
    const actualError = <Error>secondResult
    expect(actualError.message).toBe("Error: Invalid or expired verification code")
  })
})
