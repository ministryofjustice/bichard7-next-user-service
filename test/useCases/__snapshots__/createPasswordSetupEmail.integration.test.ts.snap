// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSetup should generate the email subject and body to request user to setup password 1`] = `"Finish Bichard account setup"`;

exports[`AccountSetup should generate the email subject and body to request user to setup password 2`] = `
"Hi SetupForenames undefined,

  Click here to set your password:
  http://localhost:3000/login/new-password?token=DUMMY_TOKEN
  "
`;

exports[`AccountSetup should generate the email subject and body to request user to setup password 3`] = `"<h1>Finish Bichard account setup</h1><p>Hi SetupForenames ,</p><p>Click here to set your password:</p><p>http://localhost:3000/login/new-password?token=DUMMY_TOKEN</p>"`;
